### ISSUE TITLE
screed loops endlessly on malformed fq

### ISSUE THREAD

**Issue by ctb** [dib-lab/khmer#760]
e.g. on this:

    @895:1:37:17593:9954/1
    GGTTGACGGGGCTCAGGGGGCGGCTGACTCCGAGAGACAGCAGCCGCAGCTGTCGTCAGGGGATTTCCGGGGCGGAGGCCGCAGACGCGAGTGGTGGAGGGAGAAGGCCTGACG

this:

    import screed
    for record in screed.open('foo.fq'):
       print record.name

will loop endlessly.

**Comment by kdmurray91** [comment]
Apologies in advance if this is just spam.

If anyone is interested in trying to test the fastq parser further, I made [this collection of bad FASTQs](https://github.com/kdmurray91/biofuzz/tree/master/fastq) while trying to debug similar issues in a parser I wrote. Also biopython has [this monstrosity](https://github.com/biopython/biopython/tree/master/Tests/Quality) that may be useful.\

Cheers,
Kevin

P.S., If you want to use those as test cases bundled with screed, they're yours.

**Comment by ctb** [comment]
That's brilliant - thanks, @kdmurray91!

**Comment by ctb** [comment]
(See #884 too)

**Comment by proteasome** [comment]
i'm interested in working on this

**Comment by proteasome** [comment]
@mr-c @ahaerpfer looks like the problem is here:
a malformed file may never escape the while loop, thereby never return a value for the upstream iterable

screed/fastq.py(38)fastq_iter()
 35  	        # Extract the sequence lines
 36  	        sequence = []
 37  	        line = handle.readline().strip()
 38  ->	        while not line.startswith('+') and not line.startswith('#'):
 39  	            sequence.append(line)
 40  	            line = handle.readline().strip()
 41  	
 42  	        data['sequence'] = ''.join(sequence)
 
in the case of the @ctb example, line is is ''.so

**Comment by ctb** [comment]
Would be great to handle the list of examples in https://github.com/ged-lab/khmer/issues/760#issuecomment-73407832 too!

**Comment by proteasome** [comment]
- [ ] Is it mergable?
- [ ] Did it pass the tests?
- [ ] If it introduces new functionality, is it tested?
- [ ] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  and `make doc` output. Use `make format` and manual fixing as needed.
- [ ] Is it documented in the ChangeLog?
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?

**Comment by ctb** [comment]
@proteasome this is an issue - is there a pull request that you're working on somewhere?

**Comment by ahaerpfer** [comment]
@ctb @mr-c @proteasome: I have some more or less related comments here …

The while loop above simply fails to check if we have a line at all, so the quick fix for this issue to break out of the loop could be something like this:
```
-        while not line.startswith('+') and not line.startswith('#'):
+        while line and not line.startswith('+') and not line.startswith('#'):
```

However, this would still not solve some other parsing issues that exist, in particular around empty lines in the FASTQ file.  (I guess this also relates to issue #108.)

The current implementation of *fastq_iter()* behaves like this:
* Silently stops parsing the FASTQ file at a blank line between individual records … looks rather dangerous to me.
* Continues parsing after blank lines at certain positions *within* a particular record. (Implicit logic for continuation lines … ?)
* Loops infinitely if quality information is missing for a record.

The above quick fix would change the behaviour as follows:
* Now raises an IOError if quality information is missing.
* Still silently stops parsing at blank lines between records. :-(
* Chokes on all other empty lines with an IOError.

One thing that complicates matters is that we immediately throw away the newline character after reading a line from the stream:

        line = handle.readline().strip()

This way we can't reliably discern between just an empty line and EOF anymore (see https://docs.python.org/2/tutorial/inputoutput.html#methods-of-file-objects).

I'd volunteer to refactor *fastq_iter()* so that the behaviour is more consistent, incuding sensible error messages.  (Counting input lines and mentioning the line number in error messages might also help users). [Might take until after I'm back home from the conference.]

Open question here:
* How to treat empty lines?  (I'd go with @mr-c's comment in #108 to be liberal in what we accept and simply discard them.)
* Do you actually need a mechanism for continuation lines for bases and quality info?  If so this should probably be more explicit in the code, if not, the parser should also throw an error.

In the meanwhile if you want to go for the quick fix, I'll happily open a pull request for this [@proteasome has already left PyCon; don't know when he'll be online again].

Opinions, complaints, hints, pizza, … ?

**Comment by ctb** [comment]
Two thoughts --

first, we should work towards having tests in khmer for tricky fastq parsing for both the seqan (C++) and screed (Python) parsers; it's probably important to make sure they don't barf on too many different things.

second, I would be in favor of minimalistic changes to the existing parser unless we run into widespread real-world situations that indicate further problems.

**Comment by kdmurray91** [comment]
If you want, I'll add a verbose C++ test driver and a series of tests to the jenkins build script. This way we can clone the broken fastq pool only on jenkins and not lump it into the khmer repo.

**Comment by ctb** [comment]
On Fri, Apr 17, 2015 at 05:20:51AM -0700, Kevin Murray wrote:
> If you want, I'll add a verbose C++ test driver and a series of tests to the jenkins build script. This way we can clone the broken fastq pool only on jenkins and not lump it into the khmer repo.

More generally, I think we should be exposing the C++ parsing code via Python
and then we can simply do comparisons to make sure they return reasonably
consonant results... @camillescott, thoughts?

**Comment by ctb** [comment]
Specific issue fixed in https://github.com/dib-lab/screed/pull/38 but more discussion may be needed :).

**Comment by ctb** [comment]
This is fixed, no? Double-check for 2.0.

**Comment by ctb** [comment]
Fixed by dib-lab/screed#38; more discussion in #976.
